rollout/analysis.go

```diff
@@ -66,45 +72,74 @@ func (c *RolloutController) reconcileAnalysisRuns(rollout *v1alpha1.Rollout, cur
+		}
+		return currentAr, err
 	}
-
 	return currentAr, nil
 }
 
-func (c *RolloutController) createStepBasedAnalysisRun(rollout *v1alpha1.Rollout, index int32, rolloutAnalysisStep *v1alpha1.RolloutAnalysisStep, stableRS, newRS *appsv1.ReplicaSet) (*v1alpha1.AnalysisRun, error) {
-	podHash := controller.ComputeHash(&rollout.Spec.Template, rollout.Status.CollisionCount)
-	// Since the compute hash function is not guaranteed to be stable, we will use the podHash attached the newRS if possible.
-	if newRS != nil {
-		if newRsPodHash, ok := newRS.Labels[v1alpha1.DefaultRolloutUniqueLabelKey]; ok {
-			podHash = newRsPodHash
-		}
-	}
-	analysisRunLabels := analysisutil.StepLabels(rollout, index, podHash)
+func (c *RolloutController) createAnalysisRun(rollout *v1alpha1.Rollout, rolloutAnalysisStep *v1alpha1.RolloutAnalysisStep, stableRS, newRS *appsv1.ReplicaSet, labels map[string]string) (*v1alpha1.AnalysisRun, error) {
 	args := analysisutil.BuildArgumentsForRolloutAnalysisRun(rolloutAnalysisStep, stableRS, newRS)
-	ar, err := c.getAnalysisRunFromRollout(rollout, rolloutAnalysisStep, args, podHash, analysisRunLabels)
+	podHash := getPodHash(rollout, newRS)
+	ar, err := c.getAnalysisRunFromRollout(rollout, rolloutAnalysisStep, args, podHash, labels)
 	if err != nil {
 		return nil, err
 	}
 	ar, err = c.argoprojclientset.ArgoprojV1alpha1().AnalysisRuns(ar.Namespace).Create(ar)
 	if err != nil {
 		return nil, err
 	}
-	logutil.WithRollout(rollout).WithField(logutil.AnalysisRunKey, ar.Name).Infof("Created AnalysisRun for step '%d'", index)
 	return ar, nil
 }
 
+func (c *RolloutController) reconcileStepBasedAnalysisRun(rollout *v1alpha1.Rollout, currentArs []*v1alpha1.AnalysisRun, stableRS, newRS *appsv1.ReplicaSet) (*v1alpha1.AnalysisRun, error) {
+	step, index := replicasetutil.GetCurrentCanaryStep(rollout)
+	currentAr := analysisutil.FilterAnalysisRunsByName(currentArs, rollout.Status.Canary.CurrentStepAnalysisRun)
+	if step == nil || step.Analysis == nil || index == nil {
+		err := c.cancelAnalysisRuns(rollout, []*v1alpha1.AnalysisRun{currentAr})
+		if err != nil {
+			return nil, err
+		}
+		return nil, err
+	}
+	if currentAr == nil {
+		podHash := getPodHash(rollout, newRS)
+		stepLabels := analysisutil.StepLabels(*index, podHash)
+		currentAr, err := c.createAnalysisRun(rollout, step.Analysis, stableRS, newRS, stepLabels)
+		if err == nil {
+			logutil.WithRollout(rollout).WithField(logutil.AnalysisRunKey, currentAr.Name).Infof("Created AnalysisRun for step '%d'", index)
+		}
+		return currentAr, err
+	}
+
+	return currentAr, nil
+}
+
+func getPodHash(rollout *v1alpha1.Rollout, newRS *appsv1.ReplicaSet) string {
+	// Since the compute hash function is not guaranteed to be stable, we will use the podHash attached the newRS if possible.
+	if newRS != nil {
+		if podHash, ok := newRS.Labels[v1alpha1.DefaultRolloutUniqueLabelKey]; ok {
+			return podHash
+		}
+	}
+	return controller.ComputeHash(&rollout.Spec.Template, rollout.Status.CollisionCount)
```

Comment:
Do we ever expect to ever get to this call for controller.ComputeHash() ? I feel like it is an error to ever have newRS not have v1alpha1.DefaultRolloutUniqueLabelKey. I think if we don't have the hash in the labels of the newRS, I feel like we need to error so that this problem gets surfaced, otherwise we will be masking the problem.

Also we have replicasetutil.GetPodTemplateHash

Reply:
I believe this case would only occur if a user edits the new RS and removes the label. I think it should be fine to just use the replicasetutil.GetPodTemplateHash and report an error if it's not there